compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt

```diff
@@ -0,0 +1,166 @@
+        // We're looking for members/entries/companion object in an enum class or members in companion object of an enum class.
+        val calleeIsInsideEnum = calleeContainingClass.isEnumClass
+        val calleeIsInsideEnumCompanion =
+            calleeContainingClass.isCompanion && (calleeContainingClass.outerClass(context) as? FirRegularClass)?.isEnumClass == true
+        if (!calleeIsInsideEnum && !calleeIsInsideEnumCompanion) return
+
+        val enumClass =
+            if (calleeIsInsideEnum) calleeContainingClass
+            else calleeContainingClass.outerClass(context) as? FirRegularClass ?: return
+
+        val accessedContext = context.containingDeclarations.lastOrNull {
+            // To not raise an error for an access from another enum class, e.g.,
+            //   enum class EnumCompanion3(...) {
+            //     INSTANCE(EnumCompanion2.foo())
+            //   }
+            // find an accessed context within the same enum class.
+            (it as? FirSymbolOwner<*>)?.getContainingClass(context) == enumClass
+        } ?: return
+
+        val enumMemberProperties = enumClass.declarations.filterIsInstance<FirProperty>()
+        val enumEntries = enumClass.declarations.filterIsInstance<FirEnumEntry>()
+
+        // Members inside the companion object of an enum class
+        if (calleeContainingClass == enumClass.companionObject) {
+            // Uninitialized from the point of view of members or enum entries of that enum class
+            if (accessedContext in enumMemberProperties || accessedContext in enumEntries) {
+                if (calleeDeclaration is FirProperty) {
+                    // From KT-11769
+                    // enum class Fruit(...) {
+                    //   APPLE(...);
+                    //   companion object {
+                    //     val common = ...
+                    //   }
+                    //   val score = ... <!>common<!>
+                    // }
+                    reporter.reportOn(source, FirErrors.UNINITIALIZED_VARIABLE, calleeDeclaration.symbol, context)
+                } else {
+                    // enum class EnumCompanion2(...) {
+                    //   INSTANCE(<!>foo<!>())
+                    //   companion object {
+                    //     fun foo = ...
+                    //   }
+                    // }
+                    // <!>Companion<!>.foo() v.s. <!>foo<!>()
+                    if ((expression.explicitReceiver as? FirResolvedQualifier)?.symbol?.fir == enumClass.companionObject) {
+                        reporter.reportOn(
+                            expression.explicitReceiver!!.source,
+                            FirErrors.UNINITIALIZED_ENUM_COMPANION,
+                            enumClass.symbol,
+                            context
+                        )
+                    } else {
+                        reporter.reportOn(source, FirErrors.UNINITIALIZED_ENUM_COMPANION, enumClass.symbol, context)
```

Comment:
Isn't the source pointing to the entire call in this case? Should it match the declared element type (`KtSimpleNameExpression`) of `UNINITIALIZED_ENUM_COMPANION`?

Reply:
Yes, it will point to the entire call, and thus I'm using `PositioningStrategy.REFERENCE_BY_QUALIFIED` for this diagnostic in order to nail down to reference expression.

Reply:
The reported element type here must match that declared in `FirErrors`. Otherwise, you would get the following class cast exception in the IDE after https://github.com/JetBrains/kotlin/pull/4250 is merged.

```
java.lang.ClassCastException: org.jetbrains.kotlin.psi.KtCallExpression cannot be cast to org.jetbrains.kotlin.psi.KtSimpleNameExpression
	at org.jetbrains.kotlin.idea.frontend.api.fir.diagnostics.UninitializedEnumCompanionImpl.getPsi(KtFirDiagnosticsImpl.kt:1520)
	at org.jetbrains.kotlin.idea.frontend.api.fir.diagnostics.UninitializedEnumCompanionImpl.getPsi(KtFirDiagnosticsImpl.kt:1520)
	at org.jetbrains.kotlin.idea.fir.highlighter.KotlinHighLevelDiagnosticHighlightingPass.addDiagnostic(KotlinHighLevelDiagnosticHighlightingPass.kt:55)
	at org.jetbrains.kotlin.idea.fir.highlighter.KotlinHighLevelDiagnosticHighlightingPass.doCollectInformation(KotlinHighLevelDiagnosticHighlightingPass.kt:48)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:54)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$1(PassExecutorService.java:399)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1110)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$2(PassExecutorService.java:392)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:629)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:581)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.doRun(PassExecutorService.java:391)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$run$0(PassExecutorService.java:367)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:170)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:182)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:365)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:181)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

Reply:
You're right, thanks! Changed to `calleeReference`'s source, which is of `KtNameReferenceExpression`, which is a subtype of `KtSimpleNameExpression`.